soh/soh/Enhancements/crowd-control/CrowdControl.cpp

```diff
@@ -0,0 +1,524 @@
+            packet->timeRemaining = 30000;
+        }
+        else if (strcmp(packet->effectType.c_str(), "no_z") == 0) {
+            packet->effectCategory = "no_z";
+            packet->timeRemaining = 30000;
+        }
+        else if (strcmp(packet->effectType.c_str(), "spawn_wallmaster") == 0 ||
+            strcmp(packet->effectType.c_str(), "spawn_arwing") == 0 ||
+            strcmp(packet->effectType.c_str(), "spawn_darklink") == 0 ||
+            strcmp(packet->effectType.c_str(), "spawn_stalfos") == 0 ||
+            strcmp(packet->effectType.c_str(), "spawn_wolfos") == 0 ||
+            strcmp(packet->effectType.c_str(), "spawn_freezard") == 0 ||
+            strcmp(packet->effectType.c_str(), "spawn_keese") == 0 ||
+            strcmp(packet->effectType.c_str(), "spawn_icekeese") == 0 ||
+            strcmp(packet->effectType.c_str(), "spawn_firekeese") == 0 ||
+            strcmp(packet->effectType.c_str(), "spawn_tektite") == 0 ||
+            strcmp(packet->effectType.c_str(), "spawn_likelike") == 0 ||
+            strcmp(packet->effectType.c_str(), "cucco_storm") == 0) {
+            packet->effectCategory = "spawn";
+        }
+        else {
+            packet->effectCategory = "none";
+            packet->timeRemaining = 0;
+        }
+
+        // Check if running effect is possible
+        EffectResult effectResult = ExecuteEffect(packet->effectType.c_str(), packet->effectValue, 1);
+        if (effectResult == EffectResult::Retry || effectResult == EffectResult::Failure) {
+            nlohmann::json dataSend;
+            dataSend["id"] = packet->packetId;
+            dataSend["type"] = 0;
+            dataSend["timeRemaining"] = packet->timeRemaining;
+            dataSend["status"] = effectResult;
+
+            std::string jsonResponse = dataSend.dump();
+            SDLNet_TCP_Send(tcpsock, const_cast<char*> (jsonResponse.data()), jsonResponse.size() + 1);
+        } else {
+            // Check if effect already exists
+            uint8_t anotherEffect = 0;
+            for (CCPacket* pack : receivedCommands) {
+                if (pack != packet && pack->effectCategory == packet->effectCategory && pack->packetId < packet->packetId) {
+                    anotherEffect = 1;
+
+                    nlohmann::json dataSend;
+                    dataSend["id"] = packet->packetId;
+                    dataSend["type"] = 0;
+                    dataSend["timeRemaining"] = packet->timeRemaining;
+                    dataSend["status"] = EffectResult::Retry;
+
+                    std::string jsonResponse = dataSend.dump();
+                    SDLNet_TCP_Send(tcpsock, const_cast<char*> (jsonResponse.data()), jsonResponse.size() + 1);
+
+                    break;
+                }
+            }
+
+            if (anotherEffect != 1) {
+                receivedCommands.push_back(packet);
+                std::thread t = std::thread(&CrowdControl::RunCrowdControl, this, packet);
```

Comment:
Would it be possible to have a single thread which processes the receivedCommands, instead of creating a new one for each command?